pkg/asset/machines/clusterapi.go

```diff
@@ -3,41 +3,85 @@ package machines
 import (
 	"context"
 	"fmt"
+	"os"
+	"path/filepath"
 	"strings"
 
 	"github.com/pkg/errors"
 	"github.com/sirupsen/logrus"
 	v1 "k8s.io/api/core/v1"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
 	"k8s.io/utils/ptr"
 	capa "sigs.k8s.io/cluster-api-provider-aws/v2/api/v1beta2"
 	capi "sigs.k8s.io/cluster-api/api/v1beta1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/yaml"
 
 	configv1 "github.com/openshift/api/config/v1"
 	"github.com/openshift/installer/pkg/asset"
 	"github.com/openshift/installer/pkg/asset/installconfig"
 	"github.com/openshift/installer/pkg/asset/machines/aws"
 	"github.com/openshift/installer/pkg/asset/manifests/capiutils"
 	"github.com/openshift/installer/pkg/asset/rhcos"
+	"github.com/openshift/installer/pkg/clusterapi"
 	awstypes "github.com/openshift/installer/pkg/types/aws"
 	awsdefaults "github.com/openshift/installer/pkg/types/aws/defaults"
 	azuretypes "github.com/openshift/installer/pkg/types/azure"
 )
 
-// GenerateClusterAPI generates manifests for target cluster.
-func GenerateClusterAPI(ctx context.Context, installConfig *installconfig.InstallConfig, clusterID *installconfig.ClusterID, rhcosImage *rhcos.Image) (*capiutils.GenerateMachinesOutput, error) {
+var _ asset.WritableRuntimeAsset = (*ClusterAPI)(nil)
+
+// ClusterAPI is the asset for CAPI control-plane manifests.
+type ClusterAPI struct {
+	FileList []*asset.RuntimeFile
+}
+
+// Name returns a human friendly name for the operator.
+func (c *ClusterAPI) Name() string {
+	return "Cluster API Machine Manifests"
+}
+
+// Dependencies returns all of the dependencies directly needed by the
+// ClusterAPI machines asset.
+func (c *ClusterAPI) Dependencies() []asset.Asset {
+	return []asset.Asset{
+		&installconfig.InstallConfig{},
+		&installconfig.ClusterID{},
+		new(rhcos.Image),
+	}
+}
+
+// Generate generates Cluster API machine manifests.
+func (c *ClusterAPI) Generate(dependencies asset.Parents) error {
+	installConfig := &installconfig.InstallConfig{}
+	clusterID := &installconfig.ClusterID{}
+	rhcosImage := new(rhcos.Image)
+	dependencies.Get(installConfig, clusterID, rhcosImage)
+
+	// If the feature gate is not enabled, do not generate any manifests.
+	if !capiutils.IsEnabled(installConfig) {
+		return nil
+	}
+
+	if err := os.MkdirAll(filepath.Dir(capiutils.ManifestDir), 0755); err != nil {
+		return err
+	}
+
+	c.FileList = []*asset.RuntimeFile{}
+
 	var err error
-	result := &capiutils.GenerateMachinesOutput{}
 	ic := installConfig.Config
 	pool := *ic.ControlPlane
+	ctx := context.TODO()
```

Comment:
We can do this in a follow up, while for now these assets don't carry a context, it might be useful to have a `system.Context()` function that returns the CAPI system's context, which when is shutting down, should be propagate to all client calls as well

Reply:
You’re right but I think we need it in this PR. In the current state, bootstrap destroy is failing because I shutdown the controllers with `defer cancel`. I will implement your suggestion as a temporary fix until we can plumb in the context. We will also need to implement the ability to restart the local control plane for the standalone bootstrap destroy command but that can wait (and is non trivial).

Reply:
Handled in a03afe9fc3695329bc0f91fe07b3f92b4baede86